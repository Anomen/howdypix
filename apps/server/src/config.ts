// Load env variables
require("dotenv").config();

const config = {
  rabbitMq: {
    url: process.env.RABBITMQ_URL || "amqp://localhost"
  },
  serverApi: {
    baseUrl: "",
    port: parseInt(process.env.API_PORT as string) || 3004
  },
  auth: {
    code: {
      secret: "secret_code_string",
      expiry: "2h"
    },
    token: {
      secret: "secret_token_string",
      expiry: "4h"
    },
    refreshToken: {
      secret: "secret_refresh_string",
      expiry: "30d"
    }
  }
};

config.serverApi.baseUrl =
  process.env.API_BASE_URL || `http://localhost:${config.serverApi.port}`;

export default config;
