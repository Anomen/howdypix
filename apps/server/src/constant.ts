export enum EventNames {
  NEW_FILE = "NEW_FILE"
}
