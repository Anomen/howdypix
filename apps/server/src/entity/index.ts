export * from "./Source";
export * from "./Album";
export * from "./Photo";
export * from "./Search";
export * from "./SearchResult";
