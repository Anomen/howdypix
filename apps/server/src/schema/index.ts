export * from "./album";
export * from "./auth";
export * from "./user";
