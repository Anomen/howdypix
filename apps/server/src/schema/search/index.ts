export * from "./searchTypes";
export * from "./searchQuery";
