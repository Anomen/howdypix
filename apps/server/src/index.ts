import "reflect-metadata";
import config from "./config";
import { applyApolloMiddleware } from "./middleware/apollo";
import { loadUserConfig } from "./lib/loadUserConfig";
import { startFileScan } from "./lib/startFileScan";
import EventEmitter from "events";
import { Events } from "./lib/eventEmitter";
import { startRabbitMq } from "./lib/startRabbitMq";
import { startCacheDB } from "./lib/startCacheDB";
import express from "express";
import { staticHandler } from "./middleware/static";
import { emailListHandler, emailViewHandler } from "./middleware/email";
import { applyAuthMiddleware } from "./middleware/auth";

async function main() {
  const event: Events = new EventEmitter();

  const userConfig = loadUserConfig();
  console.log("User Configuration loaded:");
  console.log(userConfig);

  /**
   * Start the mandatory services
   */
  await startCacheDB(event, userConfig);
  await startRabbitMq(event, userConfig, config.rabbitMq.url);
  await startFileScan(event, userConfig);

  /**
   * Start the API
   */
  const app = express();

  app.use(express.json());

  app.get("/static/*", staticHandler);

  // Only when we develop: it's easier to check the email templates
  // and develop them.
  app.get("/email", emailListHandler);
  app.get("/email/*", emailViewHandler);

  // Authentication routes
  applyAuthMiddleware(app);

  // Attach the Apollo middlewares
  applyApolloMiddleware(app, userConfig);

  app.listen({ port: config.serverApi.port }, () => {
    console.log(`API server started on port ${config.serverApi.port}.`);
  });
}

main();
