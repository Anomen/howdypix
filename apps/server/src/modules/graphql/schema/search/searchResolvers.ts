import {
  NexusGenArgTypes,
  NexusGenFieldTypes,
  NexusGenRootTypes,
} from "@howdypix/graphql-schema/schema";
import { appDebug, generateThumbnailUrls } from "@howdypix/utils";
import { SearchResult as EntitySearchResult } from "../../../../datastore/database/entity/SearchResult";
import { appConfig } from "../../../../lib/config";
import { ApolloContext } from "../../../../types";
import { searchHelpers } from "../../helpers/searchHelpers";

const { doSearchWithCache } = searchHelpers;

const debug = appDebug("gql");

export const getSearchResolver = () => async (
  root: {},
  args: NexusGenArgTypes["Query"]["getSearch"],
  ctx: ApolloContext
): Promise<NexusGenFieldTypes["Query"]["getSearch"]> => {
  debug(`Fetching album ${args.album}.`);

  const searchResults: EntitySearchResult[] = await doSearchWithCache(
    ctx.connection,
    args
  );

  return {
    photos:
      searchResults?.map(
        (
          searchResult: EntitySearchResult
        ): NexusGenRootTypes["SearchPhoto"] => ({
          id: searchResult.photo.id.toString(),
          thumbnails: generateThumbnailUrls(appConfig.webapp.baseUrl, {
            file: searchResult.photo.file,
            dir: searchResult.photo.dir,
            source: searchResult.photo.source,
          }).map((tn) => tn.url),
          file: searchResult.photo.file,
          birthtime: Math.round(searchResult.photo.birthtime),
        })
      ) ?? [],
  };
};
