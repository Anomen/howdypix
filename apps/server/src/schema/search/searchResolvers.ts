import {
  NexusGenArgTypes,
  NexusGenFieldTypes,
  NexusGenRootTypes,
} from "@howdypix/graphql-schema/schema.d";
import { appDebug, generateThumbnailUrls } from "@howdypix/utils";
import { SearchResult as EntitySearchResult } from "../../entity/SearchResult";
import { appConfig } from "../../config";
import { ApolloContext } from "../../types.d";
import { searchHelpers } from "../../helpers/searchHelpers";

const { doSearchWithCache } = searchHelpers;

const debug = appDebug("gql");

export const getSearchResolver = () => async (
  root: {},
  args: NexusGenArgTypes["Query"]["getSearch"],
  ctx: ApolloContext
): Promise<NexusGenFieldTypes["Query"]["getSearch"]> => {
  debug(`Fetching album ${args.album}.`);

  const searchResults: EntitySearchResult[] = await doSearchWithCache(
    ctx.connection,
    args
  );

  return {
    photos:
      searchResults?.map(
        (
          searchResult: EntitySearchResult
        ): NexusGenRootTypes["SearchPhoto"] => ({
          id: searchResult.photo.id.toString(),
          thumbnails: generateThumbnailUrls(appConfig.webapp.baseUrl, {
            file: searchResult.photo.file,
            dir: searchResult.photo.dir,
            source: searchResult.photo.source,
          }).map((tn) => tn.url),
          file: searchResult.photo.file,
          birthtime: Math.round(searchResult.photo.birthtime),
        })
      ) ?? [],
  };
};
