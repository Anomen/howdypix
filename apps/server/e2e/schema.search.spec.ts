import { EntitySchema, getConnectionManager, Connection } from "typeorm";
import { SqliteConnectionOptions } from "typeorm/driver/sqlite/SqliteConnectionOptions";
import { NexusGenArgTypes } from "@howdypix/graphql-schema/schema.d";
import * as entities from "../src/entity";
import { Album, Photo, PHOTO_STATUS, Source } from "../src/entity";
// @ts-ignore
import ormConfig from "../ormconfig";
import { getSearchResolver } from "../src/schema/search/searchResolvers";

const generateSource = (
  id: number,
  overrides: Partial<Source> = {}
): Source => {
  const source = new Source();

  source.id = id;
  source.source = `source${id}`;
  source.dir = `sourceDir${id}`;

  Object.keys(overrides).forEach((key) => {
    source[key] = overrides[key];
  });

  return source;
};

const generateAlbum = (
  id: number,
  source: Source,
  overrides: Partial<Album> = {}
): Album => {
  const album = new Album();

  album.id = id;
  album.inode = `albumInode${id}`;
  album.dir = `albumDir${id}`;
  album.parentDir = `albumParentDir${id}`;
  album.sourceLk = source;
  album.source = source.source;

  Object.keys(overrides).forEach((key) => {
    album[key] = overrides[key];
  });

  return album;
};

const generatePhoto = (
  id: number,
  album: Album | null = null,
  overrides: Partial<Photo> = {}
): Photo => {
  const photo = new Photo();

  photo.id = id;
  photo.inode = parseInt(`10000${id}`, 0);
  photo.file = `File${id}`;
  photo.dir = album?.dir ?? "";
  photo.parentDir = album?.parentDir ?? "";
  photo.album = album;
  photo.source = album?.source ?? "";
  photo.status = PHOTO_STATUS.PROCESSED;
  photo.aperture = parseInt(`1111${id}`, 0);
  photo.ISO = parseInt(`2222${id}`, 0);
  photo.shutter = parseInt(`3333${id}`, 0);
  photo.size = parseInt(`4444${id}`, 0);
  photo.birthtime = new Date(parseInt(`00${id}`, 0)).getMilliseconds();
  photo.ctime = new Date(parseInt(`11${id}`, 0)).getMilliseconds();
  photo.mtime = new Date(parseInt(`22${id}`, 0)).getMilliseconds();
  photo.createDate = new Date(parseInt(`33${id}`, 0)).getMilliseconds();
  photo.make = `make${id}`;
  photo.model = `model${id}`;

  Object.keys(overrides).forEach((key) => {
    photo[key] = overrides[key];
  });

  return photo;
};

const saveAllData = (
  connection: Connection,
  {
    sources,
    albums,
    photos,
  }: {
    sources: Source[];
    albums: Album[];
    photos: Photo[];
  }
): Promise<void> =>
  Promise.all(
    sources.map((source) => connection.getRepository(Source).save(source))
  )
    .then(() =>
      Promise.all(
        albums.map((album) => connection.getRepository(Album).save(album))
      )
    )
    .then(() =>
      Promise.all(
        photos.map((photo) => connection.getRepository(Photo).save(photo))
      )
    )
    .then(() => {});

describe("search resolver", () => {
  const connectionManager = getConnectionManager();
  const connection = connectionManager.create({
    ...(ormConfig as SqliteConnectionOptions),
    database: ":memory:",
    entities: Object.keys(entities).map(
      (e): EntitySchema =>
        ((entities as unknown) as { [key: string]: EntitySchema })[e]
    ),
  });

  beforeEach(async () => {
    jest.resetAllMocks();
    jest.restoreAllMocks();

    // Reset the database
    if (connection.isConnected) {
      await connection.close();
    }

    await connection.connect();
  });

  describe("with a full database", () => {
    let sources: Source[];
    let albums: Album[];
    let photos: Photo[];

    const resetValues = (): void => {
      sources = [generateSource(0), generateSource(1), generateSource(2)];
      albums = [
        generateAlbum(0, sources[0]),
        generateAlbum(1, sources[0]),
        generateAlbum(2, sources[0]),
        generateAlbum(3, sources[1]),
      ];
      albums[2].parentDir = albums[1].dir;

      photos = [
        generatePhoto(0, albums[0], { make: "make1", model: "model1" }),
        generatePhoto(1, albums[0], { make: "make2", model: "model2" }),
        generatePhoto(2, albums[0], { make: "make2", model: "model1" }),
        generatePhoto(3, albums[1]),
        generatePhoto(4, null, {
          source: sources[0].source,
        }),
        generatePhoto(5, null, {
          source: sources[0].source,
        }),
        generatePhoto(6, albums[2], {
          parentDir: albums[1].dir,
        }),
      ];
    };

    resetValues();

    beforeEach(() => {
      resetValues();
    });

    const tests: [string, NexusGenArgTypes["Query"]["getSearch"]][] = [
      [
        "should return photos at the root with result",
        { album: "", source: sources[0].source },
      ],
      [
        "should return photos at the root without result",
        { album: "", source: sources[1].source },
      ],
      [
        "should return photos in an album with results",
        { album: photos[0].dir, source: photos[0].source },
      ],
      [
        "should return photos in an sub-album with results",
        { album: photos[6].dir, source: photos[6].source },
      ],
      [
        "should return photos with matching filters",
        {
          album: photos[0].dir,
          source: photos[0].source,
          filterBy: { make: ["make1"] },
        },
      ],
      [
        "should return photos with multiple matching filters",
        {
          album: photos[0].dir,
          source: photos[0].source,
          filterBy: { make: ["make1", "make2"] },
        },
      ],
      [
        "should return photos with different filters",
        {
          album: photos[0].dir,
          source: photos[0].source,
          filterBy: { make: ["make1"], model: ["model1"] },
        },
      ],
      [
        "should return photos with multiple different filters",
        {
          album: photos[0].dir,
          source: photos[0].source,
          filterBy: { make: ["make1", "make2"], model: ["model1"] },
        },
      ],
      [
        "should return photos with order ASC",
        {
          album: photos[0].dir,
          source: photos[0].source,
          orderBy: "DATE_ASC",
        },
      ],
      [
        "should return photos with order DESC",
        {
          album: photos[0].dir,
          source: photos[0].source,
          orderBy: "DATE_DESC",
        },
      ],
    ];

    tests.forEach(([description, args]) => {
      it(description, async () => {
        // Initialize
        await saveAllData(connection, { sources, albums, photos });

        // Run
        const results = await getSearchResolver()({}, args, {
          connection,
          user: null,
        });

        // Test
        expect(results).toMatchSnapshot();
      });
    });

    ([
      ["without filters", {}],
      ["with orderBy", { orderBy: "DATE_ASC" }],
      ["with filterBy", { filterBy: { model: "model1" } }],
    ] as [string, NexusGenArgTypes["Query"]["getSearch"]][]).forEach(
      ([description, overrides]) =>
        it(`should return the cache results if the same query is run twice ${description}`, async () => {
          // Initialize
          await saveAllData(connection, { sources, albums, photos });
          jest.spyOn(connection.getRepository(Photo), "find");

          const runSearch = async (): Promise<void> => {
            // Run
            const results = await getSearchResolver()(
              {},
              {
                album: photos[0].dir,
                source: photos[0].source,
                ...overrides,
              },
              {
                connection,
                user: null,
              }
            );

            // Test
            expect(results).toMatchSnapshot();
            expect(connection.getRepository(Photo).find).toHaveBeenCalledTimes(
              1
            );
          };

          // Run the search twice
          await runSearch();
          await runSearch();
        })
    );
  });
});
