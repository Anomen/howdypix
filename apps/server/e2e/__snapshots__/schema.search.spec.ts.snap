// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search resolver with a full database should return photos at the root with result 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 5,
      "file": "File5",
      "id": "5",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:File5x200.jpg",
        "http://localhost:3000/files/@source0:File5x600.jpg",
        "http://localhost:3000/files/@source0:File5x1280.jpg",
        "http://localhost:3000/files/@source0:File5x2048.jpg",
      ],
    },
    Object {
      "birthtime": 4,
      "file": "File4",
      "id": "4",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:File4x200.jpg",
        "http://localhost:3000/files/@source0:File4x600.jpg",
        "http://localhost:3000/files/@source0:File4x1280.jpg",
        "http://localhost:3000/files/@source0:File4x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return photos at the root without result 1`] = `
Object {
  "photos": Array [],
}
`;

exports[`search resolver with a full database should return photos in an album with results 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 2,
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
    Object {
      "birthtime": 1,
      "file": "File1",
      "id": "1",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File1x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x2048.jpg",
      ],
    },
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return photos in an sub-album with results 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 6,
      "file": "File6",
      "id": "6",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir2/File6x200.jpg",
        "http://localhost:3000/files/@source0:albumDir2/File6x600.jpg",
        "http://localhost:3000/files/@source0:albumDir2/File6x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir2/File6x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return photos with different filters 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return photos with matching filters 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return photos with multiple different filters 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 2,
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return photos with multiple matching filters 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 2,
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
    Object {
      "birthtime": 1,
      "file": "File1",
      "id": "1",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File1x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x2048.jpg",
      ],
    },
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return photos with order ASC 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
    Object {
      "birthtime": 1,
      "file": "File1",
      "id": "1",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File1x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x2048.jpg",
      ],
    },
    Object {
      "birthtime": 2,
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return photos with order DESC 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 2,
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
    Object {
      "birthtime": 1,
      "file": "File1",
      "id": "1",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File1x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x2048.jpg",
      ],
    },
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return the cache results if the same query is run twice with filterBy 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 2,
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return the cache results if the same query is run twice with filterBy 2`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 2,
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return the cache results if the same query is run twice with orderBy 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
    Object {
      "birthtime": 1,
      "file": "File1",
      "id": "1",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File1x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x2048.jpg",
      ],
    },
    Object {
      "birthtime": 2,
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return the cache results if the same query is run twice with orderBy 2`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
    Object {
      "birthtime": 1,
      "file": "File1",
      "id": "1",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File1x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x2048.jpg",
      ],
    },
    Object {
      "birthtime": 2,
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return the cache results if the same query is run twice without filters 1`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 2,
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
    Object {
      "birthtime": 1,
      "file": "File1",
      "id": "1",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File1x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x2048.jpg",
      ],
    },
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
  ],
}
`;

exports[`search resolver with a full database should return the cache results if the same query is run twice without filters 2`] = `
Object {
  "photos": Array [
    Object {
      "birthtime": 2,
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
    Object {
      "birthtime": 1,
      "file": "File1",
      "id": "1",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File1x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x2048.jpg",
      ],
    },
    Object {
      "birthtime": 0,
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
  ],
}
`;
